### PR TITLE
Ia.json

### PR DESCRIPTION
**Pull request by drewblount**
Proposed update to IA.json that addresses a couple things we've discussed lately:

1.  Algorithm configs now have (internal) names that can be referenced, like the `african_terrestrial` detector or `hotspotter_default` identifier. This makes species-specific pipelines more readable, e.g. zebras go to the african_terrestrial detector.

2. I propose the above internal names will be the reference we pass between frontend and houston, as well as the keys for names and descriptions (internationalized or otherwise) we show to users. Example of why this is nice: a humpback fin-detector call, instead of `Megaptera.novaengliae._detect_conf[1]`, will go across the pipes as `_detector.baleen_fin`. 

Houston communicates to the frontend using these algorithm keys and/or the internationalized user-facing strings. Houston translates these keys to the actual config dictionaries that are passed to Sage.

It's not a huge change to the format but does require updating each file. I propose we do those as needed rather than for MVP. Let's do an up/down vote on this change, if we reject it we'll just keep the old format and communicate between Houston and frontend with config internal paths like `Megaptera.novaengliae._detect_conf[1]`.

### CODE REVIEW

File: ia-configs/IA.zebra.json [hunk 1]
```diff
@@ -1,33 +1,42 @@
...
-            }
-          ]
-        }
+  "_README": "A value starting with '@' indicates an absolute path from the top of this config file, and parsing that @-link results in the value of that absolute path. The first half of this file (top-level _detectors and _identifiers) defines the various algorithms we use, including the configs we send to SAGE and some metadata about each config. Each algo config is under a name-key, e.g. the african_terrestrial detector.  \n The second half of the file defines the pipeline for each species by referencing the previously-defined algorithms, in the hierachy genus->species->detectors, and genus->species->iaClass->identifiers. The values of these species configurations should be @-links but literals are also supported by the parser.",
+  "_detectors": {
+    "african_terrestrial": {
+      "name": "African terrestrial mammal detector",
+      "description": "Trained on zebras, giraffes, lions, hyenas, leopards, cheetahs, and wild dogs.",
+      "config_dict": {
+        "start_detect": "https://kaiju.dyn.wildme.io:5010/api/engine/detect/cnn/lightnet/",
```
Comment (hwindsor) [inline ia-configs/IA.zebra.json:8]: NotAnExpertOnThis but surely the hostname shouldn't be specified in the config file.

File: ia-configs/IA.zebra.json [hunk 2]
```diff
@@ -1,33 +1,42 @@
...
-              },
-              "description": "HotSpotter pattern-matcher"
-            }
-          ]
-        }
+  "_README": "A value starting with '@' indicates an absolute path from the top of this config file, and parsing that @-link results in the value of that absolute path. The first half of this file (top-level _detectors and _identifiers) defines the various algorithms we use, including the configs we send to SAGE and some metadata about each config. Each algo config is under a name-key, e.g. the african_terrestrial detector.  \n The second half of the file defines the pipeline for each species by referencing the previously-defined algorithms, in the hierachy genus->species->detectors, and genus->species->iaClass->identifiers. The values of these species configurations should be @-links but literals are also supported by the parser.",
+  "_detectors": {
+    "african_terrestrial": {
+      "name": "African terrestrial mammal detector",
+      "description": "Trained on zebras, giraffes, lions, hyenas, leopards, cheetahs, and wild dogs.",
```
Comment (hwindsor) [inline ia-configs/IA.zebra.json:6]: if it supports all these animals I guess IA.zebra.json needs a new name

**Review by hwindsor — CHANGES_REQUESTED** [review]
I like the concept a lot. 
Should the name passed to houston be "african_terrestrial" or "zebra.african_terrestrial". 
I'd prefer the latter as it does not mandate that these new detector aliases need to be unique across all ia.json files.
I guess there will be some python to extract the config to send to Sage for this detection. 
The Request changes is that you have also pulled in the changes in PR173 for this and they're not relevant. I guess that when you created the git branch, you did so from my branch, not main. 
Once that PR goes into main, you should be  able to just rebase and then it'll just be the ia.json file changes in this PR.

**Review by hwindsor — APPROVED** [review]
I like the concept a lot.
Should the name passed to houston be "african_terrestrial" or "zebra.african_terrestrial".
I'd prefer the latter as it does not mandate that these new detector aliases need to be unique across all ia.json files.
I guess there will be some python to extract the config to send to Sage for this detection.
Looks like it needs a rebase which apparently doesn't work from the /rebase command.
Then all the asset_group stuff should go

File: ia-configs/IA.zebra.json (hunk 1 above)
Comment (naknomum) [reply ia-configs/IA.zebra.json:8]: well, this is the way _old world_ does things.   and it raises an interesting question:  how will we override the "default" behavior of _"use your own [docker] acm/wbia"_ ?   it seems to me that one of the goals of these .json conf files will be that they can be used *universally* on any codex.  so maybe we need to have the urls **be relative** and set the **wbia host** elsewhere in the site setup (with default being something like `http://acm` -- or whatever docker sees?

prob a good slack chat, if we dont want to derail this PR too much.

File: ia-configs/IA.zebra.json (hunk 2 above)
Comment (naknomum) [reply ia-configs/IA.zebra.json:6]: "its historical".  :)    i dont think the name of the .json file really will play into anything, right?  its just a sort of organizational level to help us group things in a way that is more digestable (for us meta-admins).   in theory, this all could be within one giant json file, conceptually.  no?

**Comment by drewblount** [comment]
@hwindsor , I envision just passing "african_terrestrial", maybe even having that same definition shared across all ia.json files (maybe we should have the algorithm definitions in a separate file(s) that are shared?). So the names would have to be unique, but in cases where that's problematic we could give them species-specific names like "zebra_detector" or "humpback_hotspotter". Could go either way though---perhaps for MVP we just call it "zebra.african_terrestrial" and don't share the algo configs across different instances.

**Comment by naknomum** [comment]
> I like the concept a lot.
> Should the name passed to houston be "african_terrestrial" or "zebra.african_terrestrial".
> I'd prefer the latter as it does not mandate that these new detector aliases need to be unique across all ia.json files.
> I guess there will be some python to extract the config to send to Sage for this detection.
> The Request changes is that you have also pulled in the changes in PR173 for this and they're not relevant. I guess that when you created the git branch, you did so from my branch, not main.
> Once that PR goes into main, you should be able to just rebase and then it'll just be the ia.json file changes in this PR.

i was going to ask this as well -- i feel like we need to prefix with the filename (as howard demonstrates) in order to "ensure" the names are unique.  e.g. we might have `hotspotter_nosv` in multiple files.  (i put _ensure_ in quotes, cuz one of us could still mistaken re-use `hotspotter_nosv` more than once in a list! so admin beware.)

likewise, i agree with python to make this "easy":   `find_json_leaf_with_id(id)` (okay, dumb name, but you get the idea)

**Comment by drewblount** [comment]
Yes, I can write the python modeled on the java IA.json parser. It's a little more complex than just reading json because of the @-link thingies. Nothing some light recursion can't accomplish!

**Review by naknomum — APPROVED** [review]
seems good; drew and i discussed that there should be a **test** which ensures that the leaf-node id/names are unique within that file.

File: ia-configs/IA.zebra.json (hunk 1 above)
Comment (colinwkingen) [reply ia-configs/IA.zebra.json:8]: I tried to pull a bunch of this out into the global file in the same directory

File: ia-configs/IA.zebra.json (hunk 2 above)
Comment (colinwkingen) [reply ia-configs/IA.zebra.json:6]: Yeah, the files were all names IA.json and I just added the server designation, I'm sure there is a more appropriate name we can come up with.

**Comment by colinwkingen** [comment]
@drewblount I don't think my two comments require action, if you resolve the conflict I'll merge this.

File: ia-configs/IA.zebra.json (hunk 1 above)
Comment (drewblount) [reply ia-configs/IA.zebra.json]: I removed the hostname so it's just the endpoint now.

**Comment by karenc** [comment]
I think there's strange conflicts because main was merged into this branch:

Screenshot_2021-05-26 Ia json by drewblount · Pull Request #176 · WildMeOrg houston

Remember not to use the "Update branch" button, use "/rebase" instead.

**Comment by karenc** [comment]
Looks like this has already been merged in #182, closing

**Pull request closed without merging**